ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,766 @@
+use client::{BlockId, EngineClient};
+use engines::clique::util::{extract_signers, recover_creator};
+use engines::{Engine, EngineError, Seal};
+use error::{BlockError, Error};
+use ethereum_types::{Address, H64, H160, H256, U256};
+use ethkey::Signature;
+use hash::KECCAK_EMPTY_LIST_RLP;
+use itertools::Itertools;
+use lru_cache::LruCache;
+use machine::{Call, EthereumMachine};
+use parking_lot::RwLock;
+use rand::Rng;
+use super::signer::EngineSigner;
+use unexpected::{Mismatch, OutOfBounds};
+use types::BlockNumber;
+use types::header::{ExtendedHeader, Header};
+
+#[cfg(not(feature = "time_checked_add"))]
+use time_utils::CheckedSystemTime;
+
+use self::block_state::CliqueBlockState;
+use self::params::CliqueParams;
+use self::step_service::StepService;
+
+mod params;
+mod block_state;
+mod step_service;
+mod util;
+
+// TODO(niklasad1): extract tester types into a separate mod to be shared in the code base
+#[cfg(test)]
+mod tests;
+
+// Protocol constants
+/// Fixed number of extra-data prefix bytes reserved for signer vanity
+pub const VANITY_LENGTH: usize = 32;
+/// Fixed number of extra-data suffix bytes reserved for signer signature
+pub const SIGNATURE_LENGTH: usize = 65;
+/// Address length of signer
+pub const ADDRESS_LENGTH: usize = 20;
+/// Nonce value for DROP vote
+pub const NONCE_DROP_VOTE: H64 = H64([0; 8]);
+/// Nonce value for AUTH vote
+pub const NONCE_AUTH_VOTE: H64 = H64([0xff; 8]);
+/// Difficulty for INTURN block
+pub const DIFF_INTURN: U256 = U256([2, 0, 0, 0]);
+/// Difficulty for NOTURN block
+pub const DIFF_NOTURN: U256 = U256([1, 0, 0, 0]);
+/// Default empty author field value
+pub const NULL_AUTHOR: Address = H160([0x00; 20]);
+/// Default empty nonce value
+pub const NULL_NONCE: H64 = NONCE_DROP_VOTE;
+/// Default value for mixhash
+pub const NULL_MIXHASH: H256 = H256([0; 32]);
+/// Default value for uncles hash
+pub const NULL_UNCLES_HASH: H256 = KECCAK_EMPTY_LIST_RLP;
+/// Default noturn block wiggle factor defined in spec.
+pub const SIGNING_DELAY_NOTURN_MS: u64 = 500;
+
+/// How many CliqueBlockState to cache in the memory.
+pub const STATE_CACHE_NUM: usize = 128;
+
+/// Vote to add or remove the beneficiary
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
+pub enum VoteType {
+	Add,
+	Remove,
+}
+
+impl VoteType {
+	/// Try to construct a `Vote` from a nonce
+	pub fn from_nonce(nonce: H64) -> Result<Self, Error> {
+		if nonce == NONCE_AUTH_VOTE {
+			Ok(VoteType::Add)
+		} else if nonce == NONCE_DROP_VOTE {
+			Ok(VoteType::Remove)
+		} else {
+			Err(EngineError::CliqueInvalidNonce(nonce))?
+		}
+	}
+
+	/// Get the rlp encoding of the vote
+	pub fn as_rlp(&self) -> Vec<Vec<u8>> {
+		match self {
+			VoteType::Add => vec![rlp::encode(&NULL_MIXHASH), rlp::encode(&NONCE_AUTH_VOTE)],
+			VoteType::Remove => vec![rlp::encode(&NULL_MIXHASH), rlp::encode(&NONCE_DROP_VOTE)],
+		}
+	}
+}
+
+/// Clique Engine implementation
+/// block_state_by_hash -> block state indexed by header hash.
```

Comment:
This shouldn't be in the public docs

```suggestion
// block_state_by_hash -> block state indexed by header hash.
```

Reply:
https://github.com/goerli/parity-goerli/pull/82/files